Handle default_datastore for vmgroup

lipingxue opened the pull request:
This PR includes the most fix needed for #1143. For the detailed design, please refer to that issue.
**With this PR:**
1. _DEFAULT tenant will be created with two privileges (```__ALL_DS``` and ```__VM_DS```) after install.
2. "--default-datastore" is a required field for "vmgroup create" command
3. user can modify "default_datastore" through "vmgroup update" command
4. when "default_datastore" is set or modified, our code will try to create a privilege to the "default_datastore" with full access privilege if it does not exist
5. user cannot set "default_datastore" to ```__ALL_DS```, an error will be generated.
6. user cannot remove a privilege if the datastore in that privilege is the "default_datastore". An error will be generated.
7. change the vmdkops_test.py and vmdkops_admin_test.py accordingly

**Not covered in this PR,but needed to fix issue #1143 completely:**
1. upgrade related code: Without this fix, _DEFAULT tenant is created with ```default_privilege``` after install, we don't have the ```default_privilege``` any more, need to replace with two privileges ```__VM_DS``` and ```ALL_DS```
2. upgrade related code:  for any vmgroup, if the "default_datastore" is set, need to create a full access privilege to that datastore if it does not exist; and need to set the "default_datastore" to ```__VM_DS``` if it does not set since with the fix in this PR,  any "vmgroup" need set "default_datastore" field
3. The upgrade code is dependent on Shahzeb's PR, I plan to work on it after his PR is merged
4. Change the AdminCLI document accordingly


**Test Summary**

The following automated test has been modified or added for the new behavior of default_datastore handling introduced in this PR.

**vmdkops_test**

- test for default vmgroup
 1. DEFAULT vmgroup, privilege to datastore "__ALL_DS", and privilege to datastore "__VM_DS""
     are present post install, issue vmdk_ops from VM which is not owned by any vmgroup vmdk_ops 
     should succeed, and the volume create with short name will be created in the vm_datastore 
     where VM lives
  2. change the default_datastore for DEFAULT vmgroup to a real datastore,  volume create with short name should be created in the default datastore instead of VM datastore
  3. Remove privilege to "__ALL_DS" and let only privilege to  "__VM_DS" present, "create volume" with short name should succeed; while try to create a volume with full name like volume@datastore, and datastore is not the vm_datastore, the creation should fail
  4.  Remove privilege to "__VM_DS" and now no privilege available for _DEFAULT vmgroup, "create volume" should fail
  5. REMOVE DEFAULT vmgroup, "create volume" should fail

- test for named vmgroup
1. named vmgroup "tenant1" is created with default_datastore set to "VM_DS", test volume
create with short name and full name.  With short name, volume is created in the default_datastore which is the vm_datastore. Also test volume remove/attach/detach/ls.
 2. change the default_datastore to from "VM_DS" to a real datastore, and set the "volume_maxsize" an "usage_quota" to test volume create with different sizes to make sure
volume create can only succeed with the "volume_maxsize" and "usage_quota" limit.
 3. rename the "tenant11" to "new_tenant1", after that, test basic volume create/attach/detach/remove/ls

- test for vmdkops on two different vmgroups
1. Two vmgroups are created with "default_datastore" set to "__VM_DS"
2. For each vmgroup, update the "default_datastore" to a real datastore (self.datastore), and try to update the privilege to this datastore, test volume create with different sizes . Run volume ls to make sure each vmgroup can only see the volumes belong to that vmgroup

- test for default_datastore
1. Named vmgroup "tenant1" is created with "default_datastore" set to "VM_DS"
2. Test volume create with short name succeeds and the volume is created in the vm_datastore
3. Modify the privilege "VM_DS" to disallow create, then volume create with short name failed
4. set the "default_datastore" to a real datastore, make sure volume create succeeds and volume is created on the "default_datastore"
 5. modify the "default_datastore" to a different real datastore, make sure volume create succeeds and volume is created on the new "default_datastore"

**AminCLI test for "vmgroup" management**

- vmgroup create
1. “vmgroup create” without  “—default-datastore” failed
2. “vmgroup create” with –default-datastore=datastore1 succeed
3. “vmgroup create” with –default-datastore=__VM_DS succeed
4. “vmgroup create” with –default-datastore=__ALL_DS failed
5. “vmgroup create” with –default-datastore=invalid_datastore_name failed
 
 

- vmgroup update
1. “vmgroup update” with --default-datastore=valid_datastore_name succeeds
2. “vmgroup update” with –default-datastore=__ALL_DS failed
 

- vmgroup access add
1. add  “usage_quota” for privilege to “__ALL_DS” and “__VM_DS” is not allowed
 

- vmgroup access set
1. change  “usage_quota” for privilege to “__ALL_DS” and “__VM_DS” is not allowed
 
 

- vmgroup access rm
1. Remove privilege to “default_datastore” failed

lipingxue commented:
@msterin @pdhamdhere @pshahzeb @govint Please review it, thanks!

govint commented on esx_service/cli/vmdkops_admin.py:
```diff
@@ -916,10 +916,7 @@ def generate_tenant_ls_rows(tenant_list):
-            default_datastore = ""
-        else:
-            default_datastore = vmdk_utils.get_datastore_name(tenant.default_datastore_url)
+        default_datastore = vmdk_utils.get_datastore_name(tenant.default_datastore_url)
```
a. Does this work for the xESX scenario? Unless the datastore is shared, get_datastore_name() won't return a valid name.

b. If the tenant.default_datastore_url is VM_DS, then it should map to the datastore name for each VM in the tenant. Which need not be the same for every VM in the tenant. Although, to keep it simple, we may want to restrict VMs in a tenant to include only those that are on the same DS.

lipingxue replied:
a) I think for xESX, the configuration we support assumes the datastore is shared. Am I miss anything here?
b) I don't think so, "VM_DS" is a special datastore tag,  and I don't think it should map for each VM in the tenant since those VM may live on different datastores.

govint replied:
The code in vmdk_ops.py seems to map the datastore to the VM's datastore as expected, which is fine.

govint commented on esx_service/cli/vmdkops_admin.py:
```diff
@@ -1097,10 +1094,7 @@ def generate_tenant_access_ls_rows(privileges):
-            datastore = ""
-        else:
-            datastore = vmdk_utils.get_datastore_name(p.datastore_url)
+        datastore = vmdk_utils.get_datastore_name(p.datastore_url)
```
Same comment as above for generate_tenant_ls_rows()

lipingxue replied:
See reply above.

govint commented on esx_service/cli/vmdkops_admin_test.py:
```diff
@@ -677,6 +692,7 @@ def test_tenant_vm(self):
         # create tenant1 without adding any vms and privilege
         error_info, tenant = auth_api._tenant_create(
                                                     name=self.tenant1_name,
+                                                    default_datastore=auth_data_const.VM_DS,
```
Does the tenant_vm_ls() show the actual datastores for each VM or VM_DS? Is VM_DS preferrable vs. showing different datastores per VM?

lipingxue replied:
It should show "VM_DS".

msterin replied:
good question. I think i t should show either VM_DS or VM_DS(read_ds_name) so they know where is the DS :-)

govint commented on esx_service/cli/vmdkops_admin_test.py:
```diff
@@ -780,28 +799,63 @@ def test_tenant_access(self):
+                                                datastore=auth_data_const.VM_DS
+                                                )
+        self.assertEqual(None, error_info)
+
```
Could add a test to create vmgroup with VM_DS as the default DS and add VMs from at least two different DS'es to the vmgroup and verify volumes created from each VM are created in the respective datastores.

govint commented on esx_service/cli/vmdkops_admin_test.py:
```diff
@@ -901,48 +955,52 @@ def test_tenant_access(self):
         error_info, privileges = auth_api._tenant_access_ls(self.tenant1_name)
         self.assertEqual(None, error_info)
 
+        # now, only have a privileg to self.datastore_name
```
Nit - privilege

lipingxue replied:
Done.

govint commented on esx_service/cli/vmdkops_admin_test.py:
```diff
@@ -901,48 +955,52 @@ def test_tenant_access(self):
         datastore=self.datastore_name)
 
-        self.assertEqual(None, error_info)
+        self.assertNotEqual(None, error_info)
```
Add a test where the default datastore is changed and tenant_access_rm()  is done on the earlier DS, that should work as its no longer the default DS.

pshahzeb replied:
This check is good. But we should also keep the earlier check which is a success for access rm (after the default datastore is changed again).

lipingxue replied:
This test is available is we have two datastores. See the test in Line972 to Line979.

govint commented on esx_service/utils/auth.py:
```diff
@@ -91,34 +91,34 @@ def get_default_tenant():
-        -- privilegs: return a list of privileges for DEFAULT privilege if
-           DEFAULT tenant exists,
-           return None on failure or DEFAULT does not exist
+        -- privilegs: return a list of privileges on datastore "__ALL_DS"
```
"privileges:"

lipingxue replied:
Done.

govint commented on esx_service/utils/auth.py:
```diff
@@ -91,34 +91,34 @@ def get_default_tenant():
+    """ Get privileges on _ALL_DS by querying the auth DB.
+        For the tenant with given tenant_uuid, this privilege
+        will match any datastore which does not have an entry
+        in privileges table explicitly
```
The comment isn't clear, is this fetching privileges for datastores that "aren't in the privileges table"? Doesn't the privileges table have an entry only for datastores for which privileges are assigned?

lipingxue replied:
We do have this privilege to "__ALL_DS" in the privilege table.  For example,  for a given tenant "tenant1", user added a privilege to "datastore1" , then in the privilege table, a entry with <tenant1, datastore1> will be added. So any request from "tenant1" to ask checking authorization to "datastore1" will match this privilege. If we have the "__ALL_DS" privilege for "tenant1", in the privilege table, an entry with <tenant1, __ALL_DS> will be added. Any request from "tenant1" to ask checking authorization to any datastore other than "datastore1" will be checked against this entry.

msterin replied:
This **is** confusing and also  sounds different from what we've discussed. Can you elaborate on what do you think admin wants to do here? Is it the old "default privilege" fallback for all ? 
Maybe "Get  _ALL_DS priviledge from AuthDB. This privilege is 'catch-all' fallbacks for datastores not explicitly bound to a tenant_uuid' I am guessing here, I do not really understand he comment

govint replied:
@lipingxue, ok, so if the DB has <vmgroup-uuid, _ALL_DS_> it will match any datastore. But then will it allow creating a volume on the datastore? Say we have VMG1 and VMG2 with DS1 and DS2 as default datastores. And say VMG1 has the __ALL_DS_ specified in the privileges table, The user may issue "volume create x@DS2 from VMG1 and the volume is created on DS2. Is that correct?

May be ok, but vmgroups are to ensure user volumes are isolated from each other via vmgroups. In which case this catch-all shouldn't be allowed.

lipingxue replied:
@govint "volume create x@DS2 from VMG1 will check the privilege <VMG1, ALL_DS> to see if create is allowed or not.  However, "volume create x@DS1 from VMG2" is not allowed since no privilege is available for VMG2 on DS1. 
So I think  vmgroup isolation still works in this case.

govint replied:
Ok, my doubt was if a user in VMG1 will be allowed to create on DS2 just with the __ALL_DS privilege, without explicitly adding DS2 to VMG1. . Which seems to be allowed. 

Is it possible to update the documentation for the predefined datastore names (VM_DS and ALL_DS) and the privileges allowed for these,. And the default or typical use cases when these must be used.

lipingxue replied:
@govint Yes. We should document it. In typical use case, "ALL_DS" is created by our code only for "DEFAULT" tenant after install. We don't recommend user to add a privilege to "ALL_DS" for named tenant.

govint replied:
Ok, do we in that case check and disallow setting this privilege for named tenants. I've not checked if that is done.

govint commented on esx_service/utils/auth.py:
```diff
@@ -91,34 +91,34 @@ def get_default_tenant():
         privileges = cur.fetchone()
     except sqlite3.Error as e:
-        error_msg = "Error {0} when querying privileges table for DEFAULT privilege".format(e)
+        error_msg = "Error {0} when querying privileges table for _ALL_DS privilege for tenant_uuid {}".format(e, tenant_uuid)
```
for tenant_uuid {1}?

lipingxue replied:
Done.

govint commented on esx_service/utils/auth_api.py:
```diff
@@ -365,11 +365,68 @@ def is_vm_duplicate(vm_list):
+
+    return None
+
+def set_default_ds_and_create_privilege(tenant, default_datastore, check_existing):
```
Possibly rename as "new_default_ds"

lipingxue replied:
Do you mean rename "default_datastore" to "new_default_ds"? This function is called by "_tenant_create"(first add "default_datastore" to tenant) and "_tenant_update" (update the "default_datastore"). So it is not always the "new_default_ds".

msterin replied:
set_default_ds() is good enough. There is no need to describe all details in the name, 
especially since you put the correct info in docstring

govint commented on esx_service/utils/auth_api.py:
```diff
@@ -365,11 +365,68 @@ def is_vm_duplicate(vm_list):
+    if error_info:
+        return error_info
+
+    datastore_url = vmdk_utils.get_datastore_url(default_datastore)
```
Should check for datastore_url to be valid before using below.

lipingxue replied:
Done.

govint commented on esx_service/utils/auth_api.py:
```diff
@@ -365,11 +365,68 @@ def is_vm_duplicate(vm_list):
+            return error_info
+
+        # the "default_datastore" to be set is the same as existing "default_datastore" for this tenant
+        if datastore_url == existing_default_ds_url:
```
Should check if datastore_url is not None. get_datastore_url() can return None if validate call  returns None.

lipingxue replied:
Done in above.

govint commented on esx_service/utils/auth_api.py:
```diff
@@ -365,11 +365,68 @@ def is_vm_duplicate(vm_list):
+    # privilege to default_datastore already exist, no need to create
+    if error_info and error_info.code == ErrorCode.PRIVILEGE_ALREADY_EXIST:
+        logging.info(error_info.msg + " not overwriting the existing privilege")
+        error_info = None
```
Restore the existing default DS if this fails.

lipingxue replied:
I am not quite clear here. Could you elaborate more?

pshahzeb replied:
I think what @govint meant was, when this api is called from tenant_update, and setting default DS fails due to some other reason than PRIVILEGE_ALREADY_EXIST, we should restore the default DS that was already set.
IMO, the if setting default DS (_tenant_access_add function) fails, the default DS would still be existing one and we don't need to explicitly restore it.

govint replied:
No, _tenant_access_add() is trying to set privileges for the new default_datastore, if that fails then should restore the default datastore to the existing default datastore.

lipingxue replied:
With this fix, user cannot update the "default_datastore" through "tenant access" command. Instead, he/she need to use "tenant update --default-datastore". Internally, it calls "set_default_datastore" function. If it fails, an error will be returned. There is no "restore", if the change to new "default_datastore" fails, then the old "default_datastore" should keep unchanged.

govint commented on esx_service/utils/auth_api.py:
```diff
@@ -749,7 +813,10 @@ def _tenant_vm_replace(name, vm_list):
+    if datastore_name == auth_data_const.ALL_DS:
         return None
 
     if not vmdk_utils.validate_datastore(datastore_name):
```
Not this change, but this result can be different on different ESX hosts, depends on what datastores are accessible on each host.

lipingxue replied:
What do you mean the result can be different? Could you elaborate?

govint replied:
Validate_datastore() checks if the name is in the in-memory datastore cache. That depends on what datastores are visible on this host. But, thats not impacting this change.

govint commented on esx_service/utils/auth_api.py:
```diff
@@ -785,9 +852,8 @@ def check_privilege_parameters(privilege):
-
 @only_when_configured()
-def _tenant_access_add(name, datastore, allow_create=None, default_datastore=False,
+def _tenant_access_add(name, datastore, allow_create=None,
```
Just asking, can datastore be the default DS? Is it possible to add and remove privileges to the default DS?

lipingxue replied:
An full access privilege to "default_datastore" will be created during "_tenant_create" automatically.
So, it later user try to add the privilege to "default_datastore", it should return with "PRIVILEGE_ALREADY_EXIST".
User can modify this privilege through "_tenant_access_set".

govint commented on esx_service/utils/auth_api.py:
```diff
@@ -973,24 +1021,28 @@ def _tenant_access_rm(name, datastore):
-        error_msg = tenant.set_default_datastore(auth_mgr.conn, "")
-        if error_msg:
-            error_info = error_code.generate_error_info(ErrorCode.INTERNAL_ERROR, error_msg)
+        msg = "{} is set as default datastore".format(datastore)
```
tenant_access_add and tenant_access_set also need this check?

lipingxue replied:
Why the "create" should be "sticky" on default_ds? Any use cases here? The check in "_tenant_access_rm" just want to make sure it must have privilege to "default_datastore" for a tenant. Why should we enforce that it must allow "create" here?

govint replied:
Initially we allow the default datastore to be initialized with "create" privileges. Why would we need to allow removing the create privilege via a tenant_access_set(). The default datastore may be the only datastore in a vmgroup and allowing removal of create privilege means no user can create volumes. Why is that a useful capability to allow?

lipingxue replied:
We give the freedom to user to modify privilege based on their needs. If they don't want to do that, they should not do it. I don't think we should restrict this to not allowing user to revoke the "create" privilege to "default_datastores". @msterin @pshahzeb @shaominchen What do you think?

govint commented on esx_service/utils/auth_api.py:
```diff
@@ -973,24 +1021,28 @@ def _tenant_access_rm(name, datastore):
+        error_info = error_code.generate_error_info(ErrorCode.INTERNAL_ERROR, error_msg)
+        return error_info
+
+    return None
```
Can the functions that add a VM to a tenant also be checked to see if the VM can access any of the datastores (including the default) assigned to the tenant. At least warn the user that VM is added but no datastores are accessible. If default DS is not VM_DS, and the default DS is not visible on the local host for example (xESX scenario).

lipingxue replied:
You mean add the check in "_tenant_add_vm"? Why we need that? It is very possible that user add VM to tenant first, and then add datastore privilege to tenant.

lipingxue replied:
Yes, I agree with you. We should document this to say that if the "default_datastore" is not shared by all the host in the tenant, the the VM on the host which can not access the shared datastore cannot create volume on that shared datastore. We recommended to choose a datastore which is shared by all host to be "default_datastore".

govint commented on esx_service/utils/auth_data.py:
```diff
@@ -664,7 +663,10 @@ def new_db(self):
+        err = self.__create_all_ds_privileges_for_default_tenant()
+        if err:
+            return err
+        err = self.__create_vm_ds_privileges_for_default_tenant()
```
Is the default tenant usable in case of this error?

lipingxue replied:
If "new_db()" returns with error, then the configuration is "NotConfigured", which is in "no access control" mode. So why we need to remove the "DEFAULT tenant"?

govint replied:
Ok, but can the user run the same command again, after new_db() returns failure and would it cause any issue? In config_init() there is a check to see if a DB already exists and ignores calling new_db.

Possibly add a test to simulate error return for __create_vm_ds_privileges_for_default_tenant()  and check behavior when user retries "config init"

lipingxue replied:
I think in this case, we need to catch the error, and when we hit the error, remove the DB. @msterin  Any comments?

I think the fix for the issue you pointed out is not related to this PR, and I will file a new issue to track that.

msterin replied:
I think new_db() should remove the new db is anything failed. 
it's a good catch and a quick fix. 
I suggest adding try...finally in new_db()
```python
def new_db)self):
  err= self.__connect()
   if err:
    return err
  try:
    <the rest of existing code>
  finally:
    if err:
        self.__close()
        os.remove(self.db_path)
```

msterin replied:
where is the issue # ?

govint commented on esx_service/utils/error_code.py:
```diff
@@ -90,11 +93,14 @@ class ErrorCode:
     ErrorCode.INIT_NEEDED: "Please init configuration with 'vmdkops_admin.py config init' before changing it.",
     ErrorCode.SQLITE3_ERROR: "Sqlite3 error - see log for more info",
+    ErrorCode.DEFAULT_DS_NAME_INVALID : "Default datastore name {} is invalid",
+    ErrorCode.DEFAULT_DS_SET_FAILED : "Set default datastore for vmgroup {} failed ({})",
```
Should this block of error codes be modified to have a uniform prefix?  The block above has PRIVILEGE_ and the one below has VMODL_ prefix to clearly identify the category of error code.

lipingxue replied:
Done.

govint commented on esx_service/vmdk_ops.py:
```diff
@@ -847,16 +848,37 @@ def executeRequest(vm_uuid, vm_name, config_path, cmd, full_vol_name, opts):
+        error_info, tenant_uuid, tenant_name = auth.authorize(vm_uuid, datastore_url, cmd, opts)
+        # "default_datastore" is set to "__VM_DS"
+        # but no privilege exist for datastore "__VM_DS"
+        # then check wheter we have a privilege to real datastore "vm_datastore", which may be
```
But isn't the default DS always given all of the default privileges (for __VM_DS)

govint commented on esx_service/vmdk_ops.py:
```diff
@@ -847,16 +848,37 @@ def executeRequest(vm_uuid, vm_name, config_path, cmd, full_vol_name, opts):
+        # but no privilege exist for datastore "__VM_DS"
+        # then check wheter we have a privilege to real datastore "vm_datastore", which may be
+        # added by user
+        if error_info == auth_data_const.NO_PRIVILEGE and datastore_url == auth_data_const.VM_DS_URL:
```
But isn't the default DS always given all of the default privileges (for __VM_DS)?

lipingxue replied:
Good catch, we don't need this line. "__VM_DS" is set to "default_datastore" and we must have privilege to "__VM_DS".

govint commented on esx_service/vmdk_ops.py:
```diff
@@ -847,16 +848,37 @@ def executeRequest(vm_uuid, vm_name, config_path, cmd, full_vol_name, opts):
+        # if the given datastore is the same as vm_datastore
+        # then we can check privilege against "__VM_DS"
+        if error_info == auth_data_const.NO_PRIVILEGE and datastore == vm_datastore:
+            error_info, tenant_uuid, tenant_name = auth.authorize(vm_uuid, vm_datastore_url, cmd, opts)
```
If the check at line 870 returned NO_PRIVILEGE then why would this succeed? If datastore == vm_datastore then datastore_url ==vm_datastore_url.

lipingxue replied:
Yes, you are right. We should check if against privilege to "__VM_DS".

govint commented on esx_service/vmdk_ops_test.py:
```diff
@@ -1008,18 +1045,19 @@ def test_vmdkops_on_tenant_vm(self):
         # set allow_create to False
         error_info = auth_api._tenant_access_set(name=self.tenant1_name,
-                                                 datastore=self.datastore_name,
+                                                 datastore=auth_data_const.VM_DS,
```
This should be disallowed, default DS privileges shouldn't be modifiable to disallow create and ls. There isn't any use case for it (yes?).

govint commented on esx_service/vmdk_ops_test.py:
```diff
@@ -1191,75 +1234,68 @@ def test_vmdkops_for_default_datastore(self):
+        # After setup(), the "default_datastore" of tenant1 has been set to "__VM_DS"
+        # A full access privilege to datastore "__VM_DS" has been created for tenant1
+        # try to create a volume, which should succeed
+        # a volume will be tried to create on the default_datastore, which is self.datastore_name
```
default DS is __VM_DS per line 1237.

lipingxue replied:
Done.

govint commented on esx_service/vmdk_ops_test.py:
```diff
@@ -1191,75 +1234,68 @@ def test_vmdkops_for_default_datastore(self):
+        self.assertEqual(None, error_info)
+        # create second volume with default size
+        # a volume will be tried to create on the default_datastore, which is self.datastore_name
+        # create should fail since on default_datastore, since access privilege to "__VM_DS" does
```
default DS is __VM_DS

govint commented on esx_service/vmdk_ops_test.py:
```diff
@@ -1191,75 +1234,68 @@ def test_vmdkops_for_default_datastore(self):
-        # create should fail since on default_datastore, it does not have "allow_create"" set to True
+        self.assertEqual(None, error_info)
+        # create second volume with default size
+        # a volume will be tried to create on the default_datastore, which is self.datastore_name
```
default DS is __VM_DS

lipingxue replied:
Done.

govint commented on esx_service/vmdk_ops_test.py:
```diff
@@ -1191,75 +1234,68 @@ def test_vmdkops_for_default_datastore(self):
-        error_info = auth_api._tenant_access_add(name=self.tenant1_name,
-                                                 datastore=self.datastore1_name,
+        error_info = auth_api._tenant_access_set(name=self.tenant1_name,
+                                                 datastore=auth_data_const.VM_DS,
```
This ability to remove the basic privilege seems unnecessary and no use case for it. Can remove and avoid these test scenarios.

lipingxue replied:
Do you mean we should disallow user to modify the privilege to "__VM_DS"? Could you elaborate more detail?

lipingxue replied:
Same as above. What you mean is that we should not allow user to modify the privilege to "__VM_DS", right? Why?

govint replied:
Consider this scenario,

1. VM group VMG1 with __VM_DS as the default.
2. User adds two VMs VM1 and VM2 on two different datastores (DS1 and DS2) to VMG1. Each VM can create volumes on their respective datastores (DS1 and DS2).
3. Now users onVM1 and VM2 are actually isolated because they use separate datastores to keep the volumes.
4. If admin changes the privileges for __VM_DS to disallow create then users on both VMs are impacted, even though they are actually on different datastores (DS1 and DS2).

I'm not sure its correct to impact all users in every VM in the VM group in this manner.  Consider VM1 and VM2 are running two independent docker service stacks. Both are impacted if the admin removed create privileges. Admin would perhaps want to revoke privileges for a particular docker service stack on one of the VMs, but there is no way to do that here.

It may be ok as an approach but the utility of it is questionable and highly impactful. At least for VM_DS as the default, we should not allow changing the privileges as it impact unrelated users.

pshahzeb commented on esx_service/cli/vmdkops_admin_test.py:
```diff
@@ -901,48 +955,52 @@ def test_tenant_access(self):
             expected_output = self.datastore_name
             self.assertEqual(expected_output, actual_output)
 
-            # add second access privilege for tenant
```
We can retain this test check. Only the default_datastore argument could be removed.

pshahzeb commented on esx_service/cli/vmdkops_admin_test.py:
```diff
@@ -901,48 +955,52 @@ def test_tenant_access(self):
             expected_output = self.datastore1_name
             self.assertEqual(expected_output, actual_output)
 
+            # swtich the "default_datastore" to self.datastore_name
```
nit: switch

pshahzeb commented on esx_service/utils/auth.py:
```diff
@@ -328,22 +328,24 @@ def check_privileges_for_command(cmd, opts, tenant_uuid, datastore_url, privileg
+            result = auth_data_const.NO_CREATE_PRIVILEGE
         if not check_max_volume_size(opts, privileges):
-            result = "volume size exceeds the max volume size limit"
+            result = auth_data_const.MAX_VOL_EXCEED
```
these constants sound more like error codes.

pshahzeb commented on esx_service/utils/auth_api.py:
```diff
@@ -399,6 +456,10 @@ def _tenant_create(name, description="", vm_list=None, privileges=None):
 
+    error_info = check_default_datastore(default_datastore)
+    if error_info:
+        return error_info
```
should return error_info, None.
Plus we need a test case to capture this error.

pshahzeb commented on esx_service/utils/auth_api.py:
```diff
@@ -449,19 +515,17 @@ def _tenant_update(name, new_name=None, description=None, default_datastore=None
+        error_info = set_default_ds_and_create_privilege(tenant=tenant,
+                                                         default_datastore=default_datastore,
+                                                         check_existing=True)
+        if error_info:
```
This function sometimes return error_info and None, while sometimes in just returns error_info.
The return should be uniform depending on the consumption of this function.

pshahzeb commented on esx_service/vmdk_ops.py:
```diff
@@ -814,16 +814,17 @@ def executeRequest(vm_uuid, vm_name, config_path, cmd, full_vol_name, opts):
+    if error_info:
+        return err(error_info.msg)
+    elif not default_datastore_url:
+        error_info = error_code.error_code_to_message[ErrorCode.DEFAULT_NOT_SET].format(tenant_name)
```
should be DEFAULT_DS_NOT_SET.
A test case for this error.

lipingxue commented:
@govint I have addressed your comments. For those that I agree, I have made the change; for those I have questions, I have replied with my question. Please take another look. Thanks!

govint commented:
Test update:

/usr/lib/vmware/vmdkops/bin/vmdkops_admin.py volume ls
Volume     Datastore   VMGroup   Capacity  Used  Filesystem  Policy  Disk Format  Attached-to  Access      Attach-as               Created By  Created Date
---------  ----------  --------  --------  ----  ----------  ------  -----------  -----------  ----------  ----------------------  ----------  ------------------------
TestVol11  bigone (1)  _DEFAULT  400MB     23MB  ext4        N/A     thin         detached     read-write  independent_persistent  2           Tue Apr 11 08:43:43 2017
TestVol00  bigone (1)  _DEFAULT  400MB     23MB  ext4        N/A     thin         detached     read-write  independent_persistent  2           Tue Apr 11 08:44:20 2017
TestVol01  bigone (1)  _DEFAULT  400MB     23MB  ext4        N/A     thin         detached     read-write  independent_persistent  2           Tue Apr 11 08:44:26 2017
TestVol02  bigone (1)  _DEFAULT  400MB     23MB  ext4        N/A     thin         detached     read-write  independent_persistent  2           Tue Apr 11 08:44:38 2017


/usr/lib/vmware/vmdkops/bin/vmdkops_admin.py vmgroup ls
Uuid                                  Name      Description                   Default_datastore  VM_list
------------------------------------  --------  ----------------------------  -----------------  -------
11111111-1111-1111-1111-111111111111  _DEFAULT  This is the default vm-group  bigone (1)

lipingxue commented:
@govint @pshahzeb I have addressed your comments. Please take a look, thanks!

msterin commented on esx_service/utils/auth_api.py:
```diff
@@ -365,11 +365,72 @@ def is_vm_duplicate(vm_list):
+    if datastore_name == auth_data_const.VM_DS:
+        return None
+    if datastore_name == auth_data_const.ALL_DS:
+        return error_code.generate_error_info(ErrorCode.DS_DEFAULT_NAME_INVALID, datastore_name)
```
FWIW, the length of line making string out of code is really annoying.
As a simple and partial fix, can you change 'from ErrCode import Error code' to "from ErrorCode import ErrorCode, generate_code_error, error_code_to_message' (or simply 'from ErrorCode import *' and drop error_code. prefix everywhere ?

lipingxue replied:
Done.

msterin commented on esx_service/utils/error_code.py:
```diff
@@ -90,11 +99,18 @@ class ErrorCode:
+    ErrorCode.PRIVILEGE_NO_MOUNT_PRIVILEGE : "No mount privilege",
+    ErrorCode.PRIVILEGE_NO_CREATE_PRIVILEGE : "No create privilege",
+    ErrorCode.PRIVILEGE_NO_DELETE_PRIVILEGE : "No delete privilege",
+    ErrorCode.PRIVILEGE_MAX_VOL_EXCEED : "volume size exceeds the max volume size limit",
```
Please keep the Capitalized message format consistent

lipingxue replied:
Done.

govint commented on esx_service/vmdk_ops.py:
```diff
@@ -847,16 +848,31 @@ def executeRequest(vm_uuid, vm_name, config_path, cmd, full_vol_name, opts):
     if not datastore:
         datastore_url = default_datastore_url
         datastore = default_datastore
+        use_default_ds = True
```
At line 846 shouldn't the "known datastores" be the ones assigned to the vmgroup - the default and any more?

lipingxue replied:
Why? I think the "known datastore" should be all the datastores in the host, right?

govint replied:
Because the user thinks they can create volumes on all the "known datastores" and when they try they get an error saying they have no privilege. Best show only those datastores that are actually "accessible to the tenant".

govint commented on esx_service/utils/auth_api.py:
```diff
@@ -365,11 +365,72 @@ def is_vm_duplicate(vm_list):
+    if datastore_url is None:
+        error_info = error_code.generate_error_info(ErrorCode.DS_DEFAULT_NAME_INVALID, default_datastore)
+    if check_existing:
+        error_msg, existing_default_ds_url = tenant.get_default_datastore(auth_mgr.conn)
```
Is tenant_get_default_datastore() returning the datastore or the datastore URL?

lipingxue replied:
It returns datastore_url.

govint commented on esx_service/utils/auth_api.py:
```diff
@@ -365,11 +365,72 @@ def is_vm_duplicate(vm_list):
+        if datastore_url == existing_default_ds_url:
+            return None
+
+    error_msg = tenant.set_default_datastore(auth_mgr.conn, datastore_url)
```
Please log a message that "existing default datastore - <value>" is being changed to "new default datastore - <value>"

lipingxue replied:
Done.

pdhamdhere commented:
@lipingxue Please check CI. Looks like some tests are failing.

lipingxue commented:
Yes. I fixed some issue in e2e testing(run "config init --local" before start the testing),  now that test passed. Now it failed in other place. I will check it.

lipingxue commented:
Now the "make test-vm" failed because this issue, which is being fixed in #1140 @govint 
In vmdk_utils.py:get_datastore_name() the given datastore isn't found but the empty list is accessed and causes the exception.

msterin commented:
then patch the issue in your branch (e.g. by cherry-pick #1140, or manually) and make sure CI passes, as the issue can mask others ....
When CI passes, it will be a race to merge - but git handles cherry-picks tracking so it would just ignore duplicate commits on rebase or merge

msterin commented on esx_service/cli/vmdkops_admin.py:
```diff
@@ -926,11 +923,12 @@ def generate_tenant_ls_rows(tenant_list):
-        privileges=[])
+    error_info, tenant = auth_api._tenant_create(name=args.name,
+                                                 default_datastore=args.default_datastore,
+                                                 description="",
```
if it's always "", why do we need description ?

msterin commented on esx_service/cli/vmdkops_admin_test.py:
```diff
@@ -126,13 +126,27 @@ def test_policy_ls_badargs(self):
 
     def test_tenant_create(self):
-        args = self.parser.parse_args((VMGROUP + ' create --name=vmgroup1 --vm-list vm1,vm2').split())
+        args = self.parser.parse_args((VMGROUP + ' create --name=vmgroup1 --default-datastore=datastore1 --vm-list vm1,vm2').split())
```
is there a test for VM_DS ? If not, we need one

msterin commented on esx_service/utils/auth_data.py:
```diff
@@ -877,24 +879,73 @@ def __create_default_tenant(self):
         return None
 
+    def get_all_ds_privileges_dict(self):
+        """Form a dictionary with all_ds privileges used in cresting of default tenant"""
```
typo: creating

lipingxue replied:
Done.

govint commented on esx_service/utils/auth_data.py:
```diff
@@ -877,24 +879,73 @@ def __create_default_tenant(self):
+                  'usage_quota': 0}]
+
+    def get_vm_ds_privileges_dict(self):
+        """Form a dictionary with VM_DS privileges used in cresting of default tenant"""
```
"creating of"

lipingxue replied:
Done

lipingxue commented:
@msterin @govint I have addressed your comments, please review it. Thanks!

govint commented on docs/user-guide/faq.md:
````diff
@@ -133,11 +133,8 @@ Step 4: Recreate the vmgroup configuration with new name “new-vmgroup1” (ass
 
 ```
-[root@localhost:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py vmgroup create --name=new-vmgroup1  --vm-list=photon-6
+[root@localhost:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py vmgroup create --name=new-vmgroup1  --vm-list=photon-6 --default-datastore=datastore1
````
This is fine, possibly add an example of using VM_DS vs. a named datastore as well. This PR or a separate PR to update these docs.

lipingxue replied:
Done.

govint commented on docs/user-guide/faq.md:
````diff
@@ -133,11 +133,8 @@ Step 4: Recreate the vmgroup configuration with new name “new-vmgroup1” (ass
 ```
-[root@localhost:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py vmgroup create --name=new-vmgroup1  --vm-list=photon-6
+[root@localhost:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py vmgroup create --name=new-vmgroup1  --vm-list=photon-6 --default-datastore=datastore1
 vmgroup 'new-vmgroup1' is created.  Do not forget to run 'vmgroup vm add' and 'vmgroup access add' commands to enable access control.
````
Is this true any more given that access privileges are assigned when the vmgroup is created.

lipingxue replied:
Change it to "vmgroup '{}' is created. Do not forget to run 'vmgroup vm add' to add vm to vmgroup"

govint commented on esx_service/cli/vmdkops_admin.py:
```diff
@@ -914,7 +914,8 @@ def generate_tenant_ls_rows(tenant_list):
         description = tenant.description
-        if not tenant.default_datastore_url or tenant.default_datastore_url == auth_data_const.DEFAULT_DS_URL:
+
+        if not tenant.default_datastore_url:
```
Could add a comment why/when a vmgroup won't have a default DS.

lipingxue replied:
Done.

govint commented on esx_service/utils/auth.py:
```diff
@@ -85,40 +86,37 @@ def get_default_tenant():
-        DEFAULT privilege will match any (tenant_uuid, datastore) pair which
-        does not have an entry in privileges table explicitly
+def get_all_ds_privileges(tenant_uuid):
+    """ Get _ALL_DS priviledge from AuthDB. This privilege is 'catch-all' fallbacks for datastores not explicitly bound
```
Nit - "privilege" - its an edge though :-)

lipingxue replied:
Done.

govint commented on esx_service/utils/auth.py:
```diff
@@ -85,40 +86,37 @@ def get_default_tenant():
         privileges = cur.fetchone()
     except sqlite3.Error as e:
-        error_msg = "Error {0} when querying privileges table for DEFAULT privilege".format(e)
+        error_msg = "Error {} when querying privileges table for _ALL_DS privilege for tenant_uuid {}".format(e, tenant_uuid)
```
Nit - could say "all ds privileges" than _ALL_DS_ for easy understanding.

lipingxue replied:
Done.

govint commented on esx_service/utils/auth.py:
```diff
@@ -180,8 +178,7 @@ def get_tenant(vm_uuid):
              vm_name = vmdk_utils.get_vm_name_by_uuid(vm_uuid)
-             err_code = error_code.ErrorCode.VM_NOT_BELONG_TO_TENANT
-             err_msg = error_code.error_code_to_message[err_code].format(vm_name)
+             err_msg = error_code_to_message[ErrorCode.VM_NOT_BELONG_TO_TENANT].format(vm_name)
```
vm_name can be None?  If user has powered off the VM and removed it while the service is trying to execute this code?

lipingxue replied:
Done.

govint commented on esx_service/utils/auth_data.py:
```diff
@@ -877,24 +879,73 @@ def __create_default_tenant(self):
+        return  [{'datastore_url': auth_data_const.ALL_DS_URL,
+                  'allow_create': 1,
+                  'max_volume_size': 0,
+                  'usage_quota': 0}]
```
Should; the ALL_DS and VM_DS dicts be kept as global vs. embedded in this function and the ones below? Being global may be better.

lipingxue replied:
I think either way works, and will keep this as it is.

govint commented on esx_service/utils/auth_data.py:
```diff
@@ -903,12 +954,14 @@ def __create_default_privileges(self):
         if error_msg:
-            err = error_code_to_message[ErrorCode.TENANT_SET_ACCESS_PRIVILEGES_FAILED].format(auth_data_const.DEFAULT_TENANT, auth_data_const.DEFAULT_DS, error_msg)
+            err = error_code_to_message[ErrorCode.TENANT_SET_ACCESS_PRIVILEGES_FAILED].format(auth_data_const.DEFAULT_TENANT,
+                                                                                              auth_data_const.ALL_DS, error_msg)
```
VM_DS?

lipingxue replied:
Good catch. Fixed.

govint commented on esx_service/utils/log_config.py:
```diff
@@ -30,7 +30,7 @@
 LOG_CONFIG_FILE = "/etc/vmware/vmdkops/log_config.json"
 
-LOG_LEVEL_DEFAULT = 'INFO'
+LOG_LEVEL_DEFAULT = 'DEBUG'
```
Pls. change back to INFO before merge.

lipingxue replied:
Done.

govint commented on esx_service/vmdk_ops.py:
```diff
@@ -777,8 +779,9 @@ def datastore_path_exist(datastore_name):
+    logging.debug("get_datastore_name: datastore_url=%s", datastore_url)
     datastore_name = vmdk_utils.get_datastore_name(datastore_url)
-    if datastore_name is None or not datastore_path_exist(datastore_name):
+    if not datastore_path_exist(datastore_name):
```
is it guaranteed that datastore_name is non-None. Because datastore_path_exist() expects that.

lipingxue replied:
Fix the code.

govint commented on esx_service/vmdk_ops.py:
```diff
@@ -819,17 +823,17 @@ def executeRequest(vm_uuid, vm_name, config_path, cmd, full_vol_name, opts):
+    if error_info:
+        return err(error_info.msg)
+    elif not default_datastore_url:
+        err_msg = error_code_to_message[ErrorCode.DS_DEFAULT_NOT_SET].format(tenant_name)
```
Isn't there a default datastore always for a tenant now?

lipingxue replied:
Yes. When no DB corruption, Line 830-831 will not execute.

govint commented on esx_service/vmdk_ops.py:
```diff
@@ -787,6 +790,7 @@ def get_datastore_name(datastore_url):
                       datastore_name)
```
Should be logging.debug("get_datastore_name: path to /vmfs/volumes/%s does not exist", datastore_name)

lipingxue replied:
Not always. datastore_name can be None, or a stale one. Therefore, /vmfs/volumes/datastore_name not always exist.  If it does not exist, we refresh the datastore cache to get the updated "datastore_name".

govint commented on esx_service/vmdk_ops.py:
```diff
@@ -844,24 +848,48 @@ def executeRequest(vm_uuid, vm_name, config_path, cmd, full_vol_name, opts):
+        # set datastore_url to a real datastore_url
+        # createVMDK() and removeVMDK() need to pass in
+        # a real datastore_url instead of url of __VM_DS
+        datastore_url = vm_datastore_url
```
Please move lines 870 - 892 into a separate function like getVMRealDatastore() or getVMPhysicalDataStore() and call that at line 860. These lines really have no place inside executeRequest().

lipingxue replied:
I put the logic for authorize check into one function. For line 886-892, I don't think it fits to the authorize_check function, and I will keep it in the executeRequest().

govint replied:
Ok that's fine

govint commented on esx_service/vmdk_ops.py:
```diff
@@ -844,24 +848,48 @@ def executeRequest(vm_uuid, vm_name, config_path, cmd, full_vol_name, opts):
                    "Known datastores: %s.\n" \
                    "Default datastore: %s" \
                    % (datastore, ", ".join(get_datastore_names_list()), default_datastore))
 
```
Again, datastore list should be whats available for the tenant. Is it possible for a VM in a tenant to be able to create volumes on any datastore all the time? Or is it possible only if ALL_DS is set for the tenant.

lipingxue replied:
No, unless ALL_DS is set for the tenant. For DEFAULT tenant,  since ALL_DS is set post install, so it will be able to create volumes on any datastore to make sure it backward compatible. For named tenant, privilege to ALL_DS is not created unless user add this privilege explicitly.

govint replied:
Ok, if the vmgroup here is the default then it's fine otherwise it should list what's allowed for the vmgroup

lipingxue replied:
Not really. The known datastore is the available datastore list on the host, not only datastores which is allowed for the specific vmgroup. For different vmgroup, the datastores which is allowed is different.

lipingxue commented:
The following automated test has been modified or added for the new behavior of default_datastore handling introduced in this PR.

**vmdkops_test**

- test for default vmgroup
 1. DEFAULT vmgroup, privilege to datastore "__ALL_DS", and privilege to datastore "__VM_DS""
     are present post install, issue vmdk_ops from VM which is not owned by any vmgroup vmdk_ops 
     should succeed, and the volume create with short name will be created in the vm_datastore 
     where VM lives
  2. change the default_datastore for DEFAULT vmgroup to a real datastore,  volume create with short name should be created in the default datastore instead of VM datastore
  3. Remove privilege to "__ALL_DS" and let only privilege to  "__VM_DS" present, "create volume" with short name should succeed; while try to create a volume with full name like volume@datastore, and datastore is not the vm_datastore, the creation should fail
  4.  Remove privilege to "__VM_DS" and now no privilege available for _DEFAULT vmgroup, "create volume" should fail
  5. REMOVE DEFAULT vmgroup, "create volume" should fail

- test for named vmgroup
1. named vmgroup "tenant1" is created with default_datastore set to "VM_DS", test volume
create with short name and full name.  With short name, volume is created in the default_datastore which is the vm_datastore. Also test volume remove/attach/detach/ls.
 2. change the default_datastore to from "VM_DS" to a real datastore, and set the "volume_maxsize" an "usage_quota" to test volume create with different sizes to make sure
volume create can only succeed with the "volume_maxsize" and "usage_quota" limit.
 3. rename the "tenant11" to "new_tenant1", after that, test basic volume create/attach/detach/remove/ls

- test for vmdkops on two different vmgroups
1. Two vmgroups are created with "default_datastore" set to "__VM_DS"
2. For each vmgroup, update the "default_datastore" to a real datastore (self.datastore), and try to update the privilege to this datastore, test volume create with different sizes . Run volume ls to make sure each vmgroup can only see the volumes belong to that vmgroup

- test for default_datastore
1. Named vmgroup "tenant1" is created with "default_datastore" set to "VM_DS"
2. Test volume create with short name succeeds and the volume is created in the vm_datastore
3. Modify the privilege "VM_DS" to disallow create, then volume create with short name failed
4. set the "default_datastore" to a real datastore, make sure volume create succeeds and volume is created on the "default_datastore"
 5. modify the "default_datastore" to a different real datastore, make sure volume create succeeds and volume is created on the new "default_datastore"

AminCLI test for "vmgroup" management

- vmgroup create
1. “vmgroup create” without  “—default-datastore” failed
2. “vmgroup create” with –default-datastore=datastore1 succeed
3. “vmgroup create” with –default-datastore=__VM_DS succeed
4. “vmgroup create” with –default-datastore=__ALL_DS failed
5. “vmgroup create” with –default-datastore=invalid_datastore_name failed
 
 

- vmgroup update
1. “vmgroup update” with --default-datastore=valid_datastore_name succeeds
2. “vmgroup update” with –default-datastore=__ALL_DS failed
 

- vmgroup access add
1. add  “usage_quota” to a value other than unlimit for privilege to “__ALL_DS” and “__VM_DS” is not allowed
 

- vmgroup access set
1. change  “usage_quota” to a value other than unlimit for privilege to “__ALL_DS” and “__VM_DS” is not allowed
 
 

- vmgroup access rm
1. Remove privilege to “default_datastore” failed

lipingxue commented:
@msterin I have put a summary of the automated test and clean up the commit. Please take a look. Thanks!

shuklanirdesh82 commented on misc/scripts/deploy-tools.sh:
```diff
@@ -35,6 +35,7 @@ TMP_LOC=/tmp/$PLUGIN_NAME
 VMDK_OPS_UNITTEST=/tmp/vmdk_ops_unit*
 BUILD_LOC=$TMP_LOC/build
 PLUGIN_LOC=$TMP_LOC/plugin
+ADMIN_CLI=/usr/lib/vmware/vmdkops/bin/vmdkops_admin.py
```
Please remove this as it is not needed/used.

lipingxue replied:
Done.

pshahzeb commented on docs/user-guide/admin-cli.md:
```diff
@@ -154,22 +215,35 @@ optional arguments:
 ### Update
 Update existing vmgroup. This command allows to update "Description" and "Default_datastore" fields, or rename an existing vmgroup.
+Valid value for "Default_datastore" includes the name of valid datastores in the ESX host or a special value "__VM_DS".
+After changing the "default_datastore" for a vmgroup, a full access privilege to the new "default_datastore" will be creaated automatically, and the existing access privilege to old "default_datastore" will remain. User can remove the access privilege to old "default_datastore" if not needed using `vmgroup access rm` subcommands.
```
typo: creaated - >created

lipingxue replied:
Done.

msterin replied:
please use the test suggested above, for "create"

lipingxue replied:
Added this test.

pshahzeb commented on esx_service/cli/vmdkops_admin.py:
```diff
@@ -914,7 +914,9 @@ def generate_tenant_ls_rows(tenant_list):
         description = tenant.description
-        if not tenant.default_datastore_url or tenant.default_datastore_url == auth_data_const.DEFAULT_DS_URL:
+        # "default_datastore_url" should always be set, and cannot be empty
+        # it can only happen when DB has some corruption
```
If this is true, shouldn't an error be returned from this function instead of setting default_datastore empty?

lipingxue replied:
I think we should log an error here and also set the default_datastore to empty since we need to show something in this field.

pshahzeb replied:
True. Or maybe the safer option is to fail the command.

pshahzeb commented on esx_service/cli/vmdkops_admin.py:
```diff
@@ -1099,7 +1100,7 @@ def generate_tenant_access_ls_rows(privileges):
     rows = []
     for p in privileges:
-        if not p.datastore_url or p.datastore_url == auth_data_const.DEFAULT_DS_URL:
+        if not p.datastore_url:
```
Similar to comment above. Can this happen? If yes, it is an error.

msterin commented on docs/user-guide/admin-cli.md:
````diff
@@ -87,45 +87,106 @@ optional arguments:
+11111111-1111-1111-1111-111111111111  _DEFAULT  This is a default vmgroup  __VM_DS
+```
+
+The "Default_datastore" field is set to "__VM_DS" for "_DEFAULT" vmgroup. Any volume create from VM which belongs to "_DEFAULT" vmgroup will be created on the datastore where VM resides.
````
we need to use the same amount of underscores in all special places. I guess we already force single underscore for _DEFAULT, so we should stick with single underscore for the rest

lipingxue replied:
Done.

msterin commented on docs/user-guide/admin-cli.md:
```diff
@@ -87,45 +87,106 @@ optional arguments:
+
+The "Default_datastore" field is set to "__VM_DS" for "_DEFAULT" vmgroup. Any volume create from VM which belongs to "_DEFAULT" vmgroup will be created on the datastore where VM resides.
+
+Two full access privileges will be created automatically post install.
```
This line is hard to understand. Better to say "When configuration is initialized with 'config init',  the access to __ALL_DS and __VM_DS for all VMs is automatically enabled"

lipingxue replied:
Done.

msterin commented on docs/user-guide/admin-cli.md:
```diff
@@ -87,45 +87,106 @@ optional arguments:
 
 Creates a new named vmgroup and optionally assigns VMs. Valid vmgroup name is only allowed to be "[a-zA-Z0-9_][a-zA-Z0-9_.-]*"
 
+"Default_datastore" is a required parameter when creating a new named vmgroup. Valid value for this field includes the name of
```
Drop "when creating a new named vmgroup". You are in "vmgroup create" :-).
Also drop "for this field". Better to say "The value is either a valid datastore name, or special string "_VM_DS".

lipingxue replied:
Done.

msterin commented on docs/user-guide/admin-cli.md:
```diff
@@ -87,45 +87,106 @@ optional arguments:
+"Default_datastore" is a required parameter when creating a new named vmgroup. Valid value for this field includes the name of
+valid datastores in the ESX host or a special value "__VM_DS".
+After setting the "default_datastore" of a named vmgroup, a full access privilege to the "default_datastore" will be added automatically
+and the volume will be created on the "default_datastore" if using short name.
```
these 2 lines are hard to parse. Better to say "After default_datastore is set, all VMs in the group have full access to it. Also, all volumes created with short names {insert link to the short/long names article} will be placed on this datastore "

lipingxue replied:
Done.

msterin commented on docs/user-guide/admin-cli.md:
````diff
@@ -87,45 +87,106 @@ optional arguments:
+
+```
+
+The "default_datastore" cannot be set to "__ALL_DS" and an error will be generated when trying to set during vmgroup create.
````
better to say ""default_datastore" cannot be set to "__ALL_DS". An attempt to do so will generate an error"

It is not only during vmgroup create. Also during an attempt to change it. Did you test for the change ?

lipingxue replied:
Done. Yes I do test the case for update default_datastore.

msterin commented on docs/user-guide/admin-cli.md:
````diff
@@ -87,45 +87,106 @@ optional arguments:
+The "default_datastore" cannot be set to "__ALL_DS" and an error will be generated when trying to set during vmgroup create.
+```
+[root@localhost:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py vmgroup create --name=vmgroup2 --default-datastore="__ALL_DS"
+Default datastore name __ALL_DS is invalid
````
The message need to be a bit better "Cannot use __ALL_DS as default datastore. Please use specific datastore name or _VM_DS special datastore"

lipingxue replied:
Done.

msterin commented on docs/user-guide/admin-cli.md:
```diff
@@ -154,22 +215,35 @@ optional arguments:
 
 ### Update
 Update existing vmgroup. This command allows to update "Description" and "Default_datastore" fields, or rename an existing vmgroup.
+Valid value for "Default_datastore" includes the name of valid datastores in the ESX host or a special value "__VM_DS".
```
"Default_datastore" is either a  valid datastore name or a special value "__VM_DS".

lipingxue replied:
Done.

msterin commented on docs/user-guide/admin-cli.md:
```diff
@@ -315,90 +389,95 @@ optional arguments:
-when you grant first datastore access for a vmgroup.
-
+Valid value for "datastore" includes the name of valid datastores in the ESX host , special value "__VM_DS" or "_ALL_DS".
+If "datastore" is set to "__VM_DS", it will create a privilege for VM in this vmgroup to access the "vm_datastore" where VM lives.
```
"Create privilege" feels like a slang. I think it's better "When DS is set to _ALL_DS, Access to all DS is l be allowed for vms in vmgroup" 

Same for the next line

lipingxue replied:
Done.

msterin commented on docs/user-guide/admin-cli.md:
````diff
@@ -315,90 +389,95 @@ optional arguments:
 datastore2  True          500.00MB         1.00GB
 ```
 
+"-volume-totalsize" cannot be set to the value other than unlimit when add privilege for special value "__VM_DS" and "__ALL_DS".
````
what does it mean ? I am confused. Are you saying "for _VM_DS and _ALL_DS special DS names, --volume-totalzie has to be set to "umlimit", or command will fail"   ?

lipingxue replied:
I mean for _VM_DS and _ALL_DS special DS names, --volume-totalzie has to be set to "umlimit".

msterin replied:
what is "set to unlimit" ? (there is no such word).). Did you want to say "should be unlimited" ? Do we have special value "unlimit" ? (I hope we do not). What is the command to be typed ?
 I do not understand what I'd need to do to "set something to unlimit"

lipingxue replied:
Yes. I mean "it should be unlimited". We don't have special value 'unlimit'. When user add an access privilege, if "volume-totalsize" is not specified, then the default value is "unlimit".

msterin commented on docs/user-guide/admin-cli.md:
````diff
@@ -315,90 +389,95 @@ optional arguments:
+"-volume-totalsize" cannot be set to the value other than unlimit when add privilege for special value "__VM_DS" and "__ALL_DS".
+```
+[root@localhost:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py vmgroup access add --name=vmgroup1 --datastore=__VM_DS  --volume-maxsize=500MB --volume-totalsize=1GB --allow-create
+Invalid value 1024MB for setting volume-totalsize on datastore __VM_DS (Only unlimit is allowed)
````
same thing, the phrase is very rough. Please correct it. "Invalid value.  For __VM_DS the only allowed value is "unlimited"'
or something like that

lipingxue replied:
Done.

msterin commented on docs/user-guide/admin-cli.md:
```diff
@@ -315,90 +389,95 @@ optional arguments:
+[root@localhost:~]
+[root@localhost:~]
+[root@localhost:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py vmgroup access add --name=vmgroup1 --datastore=__ALL_DS  --volume-maxsize=500MB --volume-totalsize=1GB --allow-create
+Invalid value 1024MB for setting volume-totalsize on datastore __ALL_DS (Only unlimit is allowed)
```
same

lipingxue replied:
Done.

msterin commented on docs/user-guide/admin-cli.md:
```diff
@@ -433,20 +512,32 @@ optional arguments:
 
 #### Remove
 Remove access to a datastore for a vmgroup.
+An access privilege cannot be removed if the datastore in this privilege is set as "default_datastore" for the vmgroup.
```
simpler is better :-)
Removing of access privilege to "default_datastore" is not suported

lipingxue replied:
Done.

msterin commented on docs/user-guide/admin-cli.md:
```diff
@@ -433,20 +512,32 @@ optional arguments:
-[root@localhost:~]  /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py vmgroup  access rm --name=vmgroup1 --datastore=datastore1
+[root@localhost:~]
+[root@localhost:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py vmgroup access rm --name=vmgroup1 --datastore=datastore1
+Remove privilege for vmgroup1 datastore1 is not allowed (datastore1 is set as default datastore)
```
same comment. Please correct the message

lipingxue replied:
Done.

msterin commented on docs/user-guide/admin-cli.md:
```diff
@@ -472,21 +563,35 @@ Sample:
 
+[root@localhost:~]
+[root@localhost:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py vmgroup access set --name=vmgroup1 --datastore=__VM_DS --volume-totalsize=1GB
+Invalid value 1024MB for setting volume-totalsize on datastore __VM_DS (Only unlimit is allowed)
```
same

lipingxue replied:
Done

msterin commented on esx_service/utils/auth_api.py:
```diff
@@ -365,14 +366,81 @@ def is_vm_duplicate(vm_list):
 
+def check_default_datastore(datastore_name):
+    """ Check datastore with given name is a valid value for default_datastore """
+    # The valid default_datastore name are:
```
Please add "returns None for success and err message for errors"

lipingxue replied:
Done.

msterin commented on esx_service/utils/auth_api.py:
```diff
@@ -365,14 +366,81 @@ def is_vm_duplicate(vm_list):
+def set_default_ds_and_create_privilege(tenant, default_datastore, check_existing):
+    """
+        Set "default_datastore" for given tenant and create a full access privilege
+        to "default_datastore" if entry does not exist
```
what is `check_existing` ? Please add docstring

lipingxue replied:
Done.

msterin commented on esx_service/utils/auth_api.py:
```diff
@@ -365,14 +366,81 @@ def is_vm_duplicate(vm_list):
+                                    allow_create=True)
+    # privilege to default_datastore already exist, no need to create
+    if error_info and error_info.code == ErrorCode.PRIVILEGE_ALREADY_EXIST:
+        logging.info(error_info.msg + " not overwriting the existing privilege")
```
please do not use the word "priviledge" in messages. Use "access privilege" or simply "access"

lipingxue replied:
Done.

msterin commented on esx_service/utils/auth_api.py:
```diff
@@ -777,14 +857,29 @@ def check_privilege_parameters(privilege):
 
+def check_usage_quota(datastore, volume_totalsize_in_MB):
+    """
+        Check the usage_quota to be set in datastore is valid or not
```
"Check if the requested quota is valid in the given datastore"

lipingxue replied:
Done.

msterin commented on esx_service/utils/auth_api.py:
```diff
@@ -777,14 +857,29 @@ def check_privilege_parameters(privilege):
+    # usage_quota on "__VM_DS" and "__ALL_DS" should be unlimited (volume_totalsize_in_MB shoule be 0)
+    if datastore == auth_data_const.VM_DS or datastore == auth_data_const.ALL_DS:
+        if volume_totalsize_in_MB is not None and volume_totalsize_in_MB > 0:
+            err_msg = "Only unlimit is allowed"
```
the message is confusing. Also - why I cannot put a limit on _ALL_DS ? This sounds wrong... We may still keep it to merge. At any rate, please elaborate here

lipingxue replied:
I think "volume_totalsize" only make sense on a specific physical datastore, it does not make sense to have "volume_totalsize" limit on a "_VM_DS" or "_ALL_DS". In addition, when the volume is created, it must be created in a physical datastore, so any row in the "volumes table", the key is <tenant_uuid, datastore_url>, here the "datastore_url" must be a url for a physical datastore.

msterin replied:
not necessarily. For example, an admin may want to say "place your volumes wherever you want. but do not take more that 2TB total storage" 
So I do not see a reason for extra code which only blocks the feature for ALL_DS or VM_DS. Now, it this extra code saves lots of complexity somewhere else,or helps to have simpler and cleaner code, I am fine with this limit  - but I'd like to understand where are the savings

lipingxue replied:
I think the saving is the logic to calculate total volume size when checking again "volume-totalsize" for "_VM_DS" and "_ALL_DS".

lipingxue replied:
What you said is make sense, and we can support it later. Filed https://github.com/vmware/docker-volume-vsphere/issues/1185 to track this.

lipingxue commented:
@msterin I have addressed your comments, please take a look.

shuklanirdesh82 commented on esx_service/cli/vmdkops_admin.py:
```diff
@@ -974,8 +981,11 @@ def tenant_ls(args):
-    print(cli_table.create(header, rows))
+    error_info, rows = generate_tenant_ls_rows(tenant_list)
+    if error_info:
+        print error_info.msg
```
this is where CI is failing.

shuklanirdesh82 replied:
root@dhcp-192-168-31-62:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py volume ls
  File "/usr/lib/vmware/vmdkops/bin/vmdkops_admin.py", line 986
    print error_info.msg
                   ^
SyntaxError: Missing parentheses in call to 'print'
